wcfsetup/install/files/ts/WoltLabSuite/Core/Ui/Redactor/Autosave.ts

```diff
@@ -0,0 +1,363 @@
+/**
+ * Manages the autosave process storing the current editor message in the local
+ * storage to recover it on browser crash or accidental navigation.
+ *
+ * @author  Alexander Ebert
+ * @copyright  2001-2019 WoltLab GmbH
+ * @license  GNU Lesser General Public License <http://opensource.org/licenses/lgpl-license.php>
+ * @module  WoltLabSuite/Core/Ui/Redactor/Autosave
+ */
+
+import * as Core from "../../Core";
+import Devtools from "../../Devtools";
+import * as EventHandler from "../../Event/Handler";
+import * as Language from "../../Language";
+import { RedactorEditor } from "./Editor";
+import * as UiRedactorMetacode from "./Metacode";
+
+interface AutosaveMetaData {
+  [key: string]: unknown;
+}
+
+interface AutosaveContent {
+  content: string;
+  meta: AutosaveMetaData;
+  timestamp: number;
+}
+
+// time between save requests in seconds
+const _frequency = 15;
+
+class UiRedactorAutosave {
+  protected _container: HTMLElement | null = null;
+  protected _editor: RedactorEditor | null = null;
+  protected readonly _element: HTMLTextAreaElement;
+  protected _isActive = true;
+  protected _isPending = false;
+  protected readonly _key: string;
+  protected _lastMessage = "";
+  protected _metaData: AutosaveMetaData = {};
+  protected _originalMessage = "";
+  protected _restored = false;
+  protected _timer: number | null = null;
+
+  /**
+   * Initializes the autosave handler and removes outdated messages from storage.
+   *
+   * @param       {Element}       element         textarea element
+   */
+  constructor(element: HTMLTextAreaElement) {
+    this._element = element;
+    this._key = Core.getStoragePrefix() + this._element.dataset.autosave!;
+    //this._overlay = null;
+
+    this._cleanup();
+
+    // remove attribute to prevent Redactor's built-in autosave to kick in
+    delete this._element.dataset.autosave;
+
+    const form = this._element.closest("form");
+    if (form !== null) {
+      form.addEventListener("submit", this.destroy.bind(this));
+    }
+
+    // export meta data
+    EventHandler.add("com.woltlab.wcf.redactor2", `getMetaData_${this._element.id}`, (data: AutosaveMetaData) => {
+      Object.entries(this._metaData).forEach(([key, value]) => {
+        data[key] = value;
+      });
+    });
+
+    // clear editor content on reset
+    EventHandler.add("com.woltlab.wcf.redactor2", `reset_${this._element.id}`, () => this.hideOverlay());
+
+    document.addEventListener("visibilitychange", () => this._onVisibilityChange());
+  }
+
+  protected _onVisibilityChange(): void {
+    if (document.hidden) {
+      this._isActive = false;
+      this._isPending = true;
+    } else {
+      this._isActive = true;
+      this._isPending = false;
+    }
```

Comment:
```suggestion
    this._isActive = !document.hidden;
    this._isPending = document.hidden;
```
I think this is easier to understand.